Fixes to support python 3.5.1 and 64-bit libs

govint opened the pull request:
These are fixes for syntax errors with Python 3.5.1

govint commented:
Fixes #480

andrewjstone commented:
I prefer removing the space between `print` and `(` so it looks more like a standard function call. Other than that, LGTM.

pdhamdhere commented:
+1 for Andrew's suggestion otherwise LGTM.

msterin commented:
+1

govint commented:
Shall I submit this for beta or wait till after that.

On Sat, Jun 18, 2016 at 5:14 AM, Mark Sterin <notifications@github.com>
wrote:

> +1
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479#issuecomment-226906054>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseGMt7MSMO2d7_qWdU2Oxkjz4hVQDks5qMzFdgaJpZM4I4qrt>
> .
>

msterin commented:
please submit after Beta tag

govint commented:
Fixed print statements in vmdkops_admin.py.

govint commented:
Fixed loading 64-bit libvmsnapshot.so in case loading the 32-bit (default) fails.

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -31,6 +31,7 @@
 
 # VSphere lib to access ESX proprietary APIs.
 DISK_LIB = "/lib/libvmsnapshot.so"
+DISK_LIB64 = "/lib64/libvmsnapshot.so"
```
Can you elaborate on how the loading of correct versions of esx  and our Libs supposed to work end to end, ? Are  you saying in esx 6.5 the 32 bit lib is missing and you use it for determine the bitness ?

govint replied:
Both 6.0 and 6.5 have 32 and 64 bit libraries installled on them. But up to
python 3.4 there isn't an issue with using the 32-bit libraries. Python 3.5
doesn't load a 32-bit library.. It depends on what python version is
installed on the host. So, attempt to load the 32-bit lib first and if that
fails then load the 64-bit version.


On Thu, Jun 23, 2016 at 10:26 PM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/utils/kvESX.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68272042>
> :
>
> > @@ -31,6 +31,7 @@
> >
> >  # VSphere lib to access ESX proprietary APIs.
> >  DISK_LIB = "/lib/libvmsnapshot.so"
> > +DISK_LIB64 = "/lib64/libvmsnapshot.so"
>
> Can you elaborate on how the loading of correct versions of esx and our
> Libs supposed to work end to end, ? Are you saying in esx 6.5 the 32 bit
> lib is missing and you use it for determine the bitness ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/00caea071825a4f36c3479c7d75ed8ba03a003bc#r68272042>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseKUr6boNMXno2UnazDodXX7EsQyyks5qOrqzgaJpZM4I4qrt>
> .
>

msterin replied:
I am still not sure how would it help for esx_service to run. Don't we need to build our vmci lib in 32 and 64 and load the correct version? And do you plan to figure out what is the correct version based on try/except above ? And how it is tested? 
All in all, I think it is better to merge the syntax repair code (print ->print() ) and the likes - with related pycompile test. 
```python
import py_compile
py_compile.compile('my_script.py')
```

And then have a separate PR to build our vmci libs in 64, and load based on version of python or version of ESX - - with related tests on python 3.5

msterin commented:
All in all, I think it is better to merge the syntax repair code (the original change you have all LGTMs for)  , with a small pycompile test on 3.5, something like 
```python
import py_compile
py_compile.compile('my_script.py')
```

And then have a separate PR to build our vmci libs in 64, and load based on version of python or version of ESX - - with related tests on python 3.5

govint commented:
Mark, no this is not done yet. As posted in telegram json.loads() has a
problem with byte strings that we get from vmcii_get_one_op(). I'll update
with the complete changes.

For now I'm building a 64-bit vmci library to test locally.. I'll send over
all changes for review. And also update the PR.

On Fri, Jun 24, 2016 at 12:17 AM, Mark Sterin <notifications@github.com>
wrote:

> All in all, I think it is better to merge the syntax repair code (the
> original change you have all LGTMs for) , with a small pycompile test on
> 3.5, something like
>
> import py_compile
> py_compile.compile('my_script.py')
>
> And then have a separate PR to build our vmci libs in 64, and load based
> on version of python or version of ESX - - with related tests on python 3.5
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479#issuecomment-228146983>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseJPVUkfEkf3e1NZgm6QXdErikGKyks5qOtTbgaJpZM4I4qrt>
> .
>

govint commented:
Pls. don't review these changes as more changes are coming in. I'll update once those are complete.

govint commented:
Fixes as below:

1. c_char_p() no longer accepts an str, needs bytes - fix is to use <str>.encode() in all places where c_char_p() was used.
2. json.loads() doesn't accept a string, needs bytes - use "<str>.decode() in places where json.loads() was complaining (only in vmdk_ops.py). KV load and store works fine with data read from the KV file.
3. Building a 64-bit vmci_srv library and installing on ESX.
4. Using a 64-bit libvmsnapshot.so on ESX and updated disklib API arg/result list in kvESX.py to use 64-bit types depending on python version used. Choice of library (32 or 64 bits) depends on >=python3.5.1
5. Fixed print, exception statements as reported by Python.

govint commented:
CI deploy is still failing.

msterin commented:
Since there have been MANY functional changes since my last review, I withdraw my LGTM  - I need to do another review now.

pdhamdhere commented:
Hey @govint Is this PR ready for review? Otherwise, do you want to close this PR and publish when it is ready for review?

My understanding so far is we need to support following versions;
- Python 2 32-bit bundle ( < ESX 6.x)
- Python 3 64-bit bundle ( >= ESX 2016)

Build should build both of the above and deploy/run should pick appropriate binaries (32-bit vs 64-bit).

msterin commented on esx_service/Makefile:
```diff
@@ -33,19 +33,21 @@ OFFLINE_DEPOT := vmware-esx-vmdkops-$(PKG_VERSION).zip
 # Shared lib. Note that it is referred by name from vmci_srv.py
-VMCI_SRV_LIB := $(BIN)/libvmci_srv.so
+VMCI_SRV_LIB   := $(BIN)/libvmci_srv.so
+VMCI_SRV_LIB64 := $(BIN)/libvmci_srv64.so
```
should it be  something like ? 
```
VMCI_SRV_LIB_NAME   := libvmci_srv.so
VMCI_SRV_LIB32   := $(BIN)/$(VMCI_SRV_LIB_NAME)
VMCI_SRV_LIB64 := $(BIN)/$(VMCI_SRV_LIB_NAME).64
```

since you will be copying it to staging area explicitly to $(VMCI_SRV_LIB_ NAME) anyways ?

msterin commented on esx_service/Makefile:
```diff
@@ -89,25 +92,29 @@ build: $(VIB_BIN)
 	@cp $(TO_ESX_PY)  $(VMDKOPS_PY)
-	@cp $(TO_ESX_LIB) $(VMDKOPS_LIB)
+	@cp $(VMCI_SRV_LIB) $(VMDKOPS_LIB)
+	@cp $(VMCI_SRV_LIB64) $(VMDKOPS_LIB64)/libvmci_srv.so
```
please do not use the explicit name here. Define VMCI_SRV_LIB_NAME  as I suggested, or simply use $(basename)

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -22,6 +22,10 @@
+import sys
+
+# Python version 3.5.1 
+PYTHON64_VERSION = 50659824
```
if python clicks to 3.5.2, will it still work ?

govint replied:
Ok, I'm checking if the Python version is >= 3.5.1. This is lowest version
thats reporting the issues and what 6.5 is going to be shipping with
(main-stage has this).

On Tue, Jun 28, 2016 at 3:10 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/utils/kvESX.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68661402>
> :
>
> > @@ -22,6 +22,10 @@
> >          c_void_p, c_char_p, c_int32, c_bool, c_uint32, c_uint64
> >  import json
> >  import logging
> > +import sys
> > +
> > +# Python version 3.5.1
> > +PYTHON64_VERSION = 50659824
>
> if python clicks to 3.5.2, will it still work ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68661402>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseGJMhfvP8cxbzLRI3MRfrQ9zh3_wks5qQEM2gaJpZM4I4qrt>
> .
>

msterin replied:
Ah, makes sense - thanks for the comment. I missed the >= the check.

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -42,24 +47,78 @@
+        lib.DiskLib_SidecarCreate.restype = int
+        useSideCarCreate = True
+    except:
+        # do nothing
```
"do nothing" comment is incorrect as you do print a  message. 
You can always use 
```python
except:
  logging.debug(...)
  pass
```
if you want to emphasize the lack of specific action

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -42,24 +47,78 @@
+    lib.DiskLib_SidecarMakeFileName.argtypes = [c_char_p, c_char_p]
+    lib.DiskLib_SidecarMakeFileName.restype = c_char_p
+
+    lib.ObjLib_Pread.argtypes = [c_uint32, c_void_p, c_uint64, c_uint64]
```
where do we use pread/pwrite ?

govint replied:
These aren't used presently nor are the DDB get/set but in place in case we
need to support <6.0 ESX or want to write to side cars with these. Although
pread/write() is not usable (I'll remove that). Have raised a PR in
bugzilla to create disklib functions that will do generic read/write IO -
handle all of the the direct IO and buffer alignment.

On Tue, Jun 28, 2016 at 3:12 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/utils/kvESX.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68661734>
> :
>
> > +    except:
> > +        # do nothing
> > +        logging.debug(
> > +            "ESX version doesn't support create API, using open instead.")
> > +
> > +    lib.DiskLib_SidecarOpen.argtypes = [c_uint64, c_char_p, c_int32,
> > +                                        POINTER(c_uint64)]
> > +    lib.DiskLib_SidecarOpen.restype = int
> > +
> > +    lib.DiskLib_SidecarClose.argtypes = [c_uint64, c_char_p, POINTER(c_uint64)]
> > +    lib.DiskLib_SidecarClose.restype = int
> > +
> > +    lib.DiskLib_SidecarMakeFileName.argtypes = [c_char_p, c_char_p]
> > +    lib.DiskLib_SidecarMakeFileName.restype = c_char_p
> > +
> > +    lib.ObjLib_Pread.argtypes = [c_uint32, c_void_p, c_uint64, c_uint64]
>
> where do we use pread/pwrite ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68661734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseBTwOhIT3-rxERvbcZWPgmyoiwuDks5qQEPIgaJpZM4I4qrt>
> .
>

msterin replied:
let's add only what we need at the moment. We are not providing generic wrapper, we just use what is needed

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -42,24 +47,78 @@
-def kvESXInit():
+
+# Initialize 32-bit disk lib interfaces
+def kv_ESX_lib64_init():
```
why 64 bit init is different from 32 bit init? Looking into function defintion, they seem to using 32 or 64 bit explicitly, so why different functions? 

[nit] if it **is** needed, I'd prefer to hide this in kv_esx_lib_init() rather that having it done upstairs

govint replied:
Args are 64-bit handles passed to disklib in the 64-bit init and 32-bit
handles otherwise. I didn't want to have one large function with most
functions declared in a if - else statement. I can separate out the ones
that are not impacted into a kv_esx_lib_common() so that just the 32-64-bit
declarations are in one function (separated of course).

On Tue, Jun 28, 2016 at 3:14 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/utils/kvESX.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68662102>
> :
>
> >
> >
> >  # Loads the ESX library used to access proprietary ESX disk lib API.
> >  # Create arg/result definitions for those that we use.
> > -def kvESXInit():
> > +
> > +# Initialize 32-bit disk lib interfaces
> > +def kv_ESX_lib64_init():
>
> why 64 bit init is different from 32 bit init? Looking into function
> defintion, they seem to using 32 or 64 bit explicitly, so why different
> functions?
>
> [nit] if it *is* needed, I'd prefer to hide this in kv_esx_lib_init()
> rather that having it done upstairs
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68662102>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseImn0eDt8YCRVocgYyfzflLbBf0Lks5qQERXgaJpZM4I4qrt>
> .
>

msterin replied:
You can wrap 2 sets of definitions it
2 big blocks in one function - essentially 
the same way you did one KV init function

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -102,17 +161,33 @@ def kvESXInit():
-    dhandle = c_uint32(0)
-    ihandle = c_uint32(0)
+    if using_64bits:
+       dhandle = c_uint64(0)
```
Can't you just have a wrapper for c_uint (get_c_uint) and call it when needed, instead of doing ' if using_64bits:' all over the place ?

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -123,42 +198,57 @@ def volOpenPath(volpath):
 def create(volpath, kv_dict):
-    disk = c_uint32(0)
-    objHandle = c_uint32(0)
+    if using_64bits:
```
same comment - please wrap c_uint conversion and use it instead of multiple  'if using_64bits:'

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -648,7 +653,9 @@ def disk_attach(vmdk_path, vm):
                      controller_key])
         # search in 15 slots, with unit_number 7 reserved for scsi controller
-        availSlots = set(range(0, 6) + range(8, 16)) - taken
+        range1 = set(range(0, 7))
```
why do we need 3 lines instad of one now ? I understand  you fixed a bug (Missing slot #6), but why not simply leave it as 
```python
availSlots = set(range(0, 7) + range(8, 16)) - taken
``` 
?

govint replied:
Python 3.5 doesn't allow '+' in a set definition:

Python 3.5.1+ (default, Apr  4 2016, 12:10:58)
[GCC 4.6.3] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> y=set(range(1,5) + range(10,15))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'range' and 'range'


Works perfect in 2.7,
Python 2.7.6 (default, Mar 22 2014, 22:59:56)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> y=set(range(1,5) + range(10,15))
>>> print y
set([1, 2, 3, 4, 10, 11, 12, 13, 14])
>>>


On Tue, Jun 28, 2016 at 3:19 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/vmdk_ops.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68662806>
> :
>
> > @@ -648,7 +653,9 @@ def disk_attach(vmdk_path, vm):
> >                       if type(dev) == vim.VirtualDisk and dev.controllerKey ==
> >                       controller_key])
> >          # search in 15 slots, with unit_number 7 reserved for scsi controller
> > -        availSlots = set(range(0, 6) + range(8, 16)) - taken
> > +        range1 = set(range(0, 7))
>
> why do we need 3 lines instad of one now ? I understand you fixed a bug
> (Missing slot #6
> <https://github.com/vmware/docker-volume-vsphere/issues/6>), but why not
> simply leave it as
>
> availSlots = set(range(0, 7) + range(8, 16)) - taken
>
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68662806>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseFMt0FppY3eb9jvekbPcqS97DAccks5qQEV0gaJpZM4I4qrt>
> .
>

msterin replied:
OK, still no need for 3 lines
availSlots = (set(range(0, 7) | range(8, 16)) - taken

govint replied:
Nope, fails as well,
Python 3.4.0 (default, Apr 11 2014, 13:05:11)
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> y=set(range(0, 7) | range(8, 16))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'range' and 'range'


On Tue, Jun 28, 2016 at 10:52 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/vmdk_ops.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68698925>
> :
>
> > @@ -648,7 +653,9 @@ def disk_attach(vmdk_path, vm):
> >                       if type(dev) == vim.VirtualDisk and dev.controllerKey ==
> >                       controller_key])
> >          # search in 15 slots, with unit_number 7 reserved for scsi controller
> > -        availSlots = set(range(0, 6) + range(8, 16)) - taken
> > +        range1 = set(range(0, 7))
>
> OK, still no need for 3 lines
> availSlots = (set(range(0, 7) | range(8, 16)) - taken
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68698925>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseAf7a2PLZMaDpMZpcQBt1mvzoGHOks5qQK-FgaJpZM4I4qrt>
> .
>

msterin replied:
@govint - reading manual helps.
```python
availSlots = ( set(range(0,7))  | set(range(7, 15)) )  - taken
```
works fine on both pyth2 and pyth3

govint replied:
I'll do this - set(range(0, 7)) | set(range(8, 16)) - taken

On Tue, Jun 28, 2016 at 10:54 AM, Govindan T <tgovin1@gmail.com> wrote:

> Nope, fails as well,
> Python 3.4.0 (default, Apr 11 2014, 13:05:11)
> [GCC 4.8.2] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> y=set(range(0, 7) | range(8, 16))
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: unsupported operand type(s) for |: 'range' and 'range'
>
>
> On Tue, Jun 28, 2016 at 10:52 AM, Mark Sterin <notifications@github.com>
> wrote:
>
>> In esx_service/vmdk_ops.py
>> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68698925>
>> :
>>
>> > @@ -648,7 +653,9 @@ def disk_attach(vmdk_path, vm):
>> >                       if type(dev) == vim.VirtualDisk and dev.controllerKey ==
>> >                       controller_key])
>> >          # search in 15 slots, with unit_number 7 reserved for scsi controller
>> > -        availSlots = set(range(0, 6) + range(8, 16)) - taken
>> > +        range1 = set(range(0, 7))
>>
>> OK, still no need for 3 lines
>> availSlots = (set(range(0, 7) | range(8, 16)) - taken
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68698925>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/APHseAf7a2PLZMaDpMZpcQBt1mvzoGHOks5qQK-FgaJpZM4I4qrt>
>> .
>>
>
>

govint replied:
Missed the "saving lines" section :-)), anyway I got the same approach
above.

On Tue, Jun 28, 2016 at 11:07 AM, Mark Sterin <notifications@github.com>
wrote:

> In esx_service/vmdk_ops.py
> <https://github.com/vmware/docker-volume-vsphere/pull/479#discussion_r68699882>
> :
>
> > @@ -648,7 +653,9 @@ def disk_attach(vmdk_path, vm):
> >                       if type(dev) == vim.VirtualDisk and dev.controllerKey ==
> >                       controller_key])
> >          # search in 15 slots, with unit_number 7 reserved for scsi controller
> > -        availSlots = set(range(0, 6) + range(8, 16)) - taken
> > +        range1 = set(range(0, 7))
>
> @govint <https://github.com/govint> - reading manual helps.
>
> availSlots = ( set(range(0,7))  | set(range(7, 15)) )  - taken
>
> works fine on both pyth2 and pyth3
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479/files/618f13068273be9c4251cba120e77e704c061ed4#r68699882>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseBUtFrYZtbaucwWjI2ZSWEzIdspEks5qQLMqgaJpZM4I4qrt>
> .
>

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -738,13 +745,19 @@ def signal_handler_stop(signalnum, frame):
+def load_vmci():
+    global lib
+
+    try:
```
this should be based on 32/64 bit switch, not simply failing to load

govint commented:
Yes please all do the review again. There was no point fixing a few things
and then the server still doesn't even run.
On Jun 28, 2016 2:50 AM, "Mark Sterin" <notifications@github.com> wrote:

> Since there have been MANY functional changes since my last review, I
> withdraw my LGTM - I need to do another review now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479#issuecomment-228879200>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseODDVwhdfywlubW5tT73j6NNLlcYks5qQD6ugaJpZM4I4qrt>
> .
>

govint commented:
Prashant,
    I've posted the changes for review. The appropriate libs are built and
used depending on the python version used.

On Tue, Jun 28, 2016 at 3:02 AM, Prashant Dhamdhere <
notifications@github.com> wrote:

> Hey @govint <https://github.com/govint> Is this PR ready for review?
> Otherwise, do you want to close this PR and publish when it is ready for
> review?
>
> My understanding so far is we need to support following versions;
>
>    - Python 2 32-bit bundle ( < ESX 6.x)
>    - Python 3 64-bit bundle ( >= ESX 2016)
>
> Build should build both of the above and deploy/run should pick
> appropriate binaries (32-bit vs 64-bit).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/479#issuecomment-228882113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseHOQmZN8TzOGErfyWMqJSUqvJuJOks5qQEFegaJpZM4I4qrt>
> .
>

msterin commented on esx_service/Makefile:
```diff
@@ -89,25 +93,29 @@ build: $(VIB_BIN)
 	@cp $(TO_ESX_BIN) $(VMDKOPS_BIN)
 	@cp $(TO_ESX_PY)  $(VMDKOPS_PY)
-	@cp $(TO_ESX_LIB) $(VMDKOPS_LIB)
+	@cp $(VMCI_SRV_LIB) $(VMDKOPS_LIB)
```
nit: just for consistency, you may want to change this line to 
`@cp $(VMCI_SRV_LIB) $(VMDKOPS_LIB)//$(VMCI_SRV_LIB_NAME)`

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2016 VMware, Inc. All Rights Reserved.
+	# Copyright 2016 VMware, Inc. All Rights Reserved.
```
nit: no need for the extra spaces here

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -42,130 +47,156 @@
-    lib.DiskLib_SidecarClose.argtypes = [c_uint32, c_char_p, POINTER(c_uint32)]
-    lib.DiskLib_SidecarClose.restype = int
+    # Define arg types for disk lib apis.
+    if sys.hexversion >= PYTHON64_VERSION:
```
I am not sure why did you take all disklib init stuff from loadDiskLib() and moved to kv_esx_init() ? 
If it was to address my comment about spread of 32 vs 64 bit code, I think it's fine in init functions (which is also what I said). 

I think disklib-related init correct place in in loadDiskLib (or disklib_init(), using the proper convention). The current arrangement feels wrong. E..g. if I want to use DiskLib_OpenWithInfo, I for some reason need to call kv_esx_init() first, instead of just disklib_init().

govint replied:
Correct, will revert.

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -738,13 +745,19 @@ def signal_handler_stop(signalnum, frame):
+def load_vmci():
+   global lib
+
+   if sys.hexversion >= PYTHON64_VERSION:
```
if possible, please add an info log message

msterin commented:
Overall LGTM, pending one issues resolution  - the location of disklib init code seems wrong. Please review and either fix or explain.
Also a couple of small nits ( feel free to ignore).

andrewjstone commented on esx_service/utils/kvESX.py:
```diff
@@ -42,130 +47,156 @@
 
-    lib.DiskLib_SidecarClose(disk, DVOL_KEY, byref(objHandle))
-    lib.DiskLib_Close(disk)
+       lib.DiskLib_SidecarClose(dhandle, DVOL_KEY.encode(), byref(objHandle))
```
Are these indentations correct? I'm not sure they aren't, but it looks like they changed to move under
 `if disk_is_valid(dhandle):`

In this case there is no other return value when the disk is invalid and None will be returned. I'm not sure if this is correct or not since there is no specification of the return value or any doc comments explaining what is happening.

govint replied:
Thanks, my change was to put the code into the if disk valid clause while earlier if the disk was invalid the function returned there itself. Reverted.

andrewjstone commented on esx_service/utils/kvESX.py:
```diff
@@ -42,130 +47,156 @@
-
-    lib.DiskLib_Close(disk)
-    return True
+       lib.DiskLib_Close(dhandle)
```
Same issue as my comment on create. In this function however, it appears that returning None is absolutely incorrect and False should be returned if `disk_is_valid` returns false.

msterin replied:
good catch !

govint replied:
Yes, reverted the code to its earlier form and avoid moving all of the code into the "if disk valid clause". Thanks

andrewjstone commented on esx_service/vmdk_ops.py:
```diff
@@ -808,7 +821,9 @@ def handleVmciRequests(port):
                          req["cmd"], ret)
 
-        response = lib.vmci_reply(c, c_char_p(json.dumps(ret)))
+            ret_string = json.dumps(ret)
```
Should these 2 lines really be moved inside the else clause? If so, response will be None if there is an exception and we will never match `if response == VMCI_ERROR:`

govint replied:
Correct, thanks

andrewjstone commented:
Local `make all` is showing an error. Note that the same error is not seen on master.
```
Running unit tests in /tmp/vmdk_ops_unittest/vmdkops_admin_sanity_test.py...
..E
======================================================================
ERROR: test_status (__main__.TestVmdkopsAdminSanity)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/vmdk_ops_unittest/vmdkops_admin_sanity_test.py", line 58, in test_status
    output = subprocess.check_output(cmd, shell=True, stderr=self.devnull)
  File "/build/mts/release/bora-3620759/bora/build/esx/release/vmvisor/sys-boot/lib/python2.7/subprocess.py", line 573, in check_output
CalledProcessError: Command '/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py status' returned non-zero exit status 1

----------------------------------------------------------------------
Ran 3 tests in 1.876s

FAILED (errors=1)
Running unit tests in /tmp/vmdk_ops_unittest/vmdkops_admin_test.py...
.....................
----------------------------------------------------------------------
Ran 21 tests in 2.606s

OK
Running unit tests in /tmp/vmdk_ops_unittest/vol_test.py...
VOL_TEST kv_store skippped
Running unit tests in /tmp/vmdk_ops_unittest/vsan_info_test.py...
ss
----------------------------------------------------------------------
Ran 2 tests in 0.000s

OK (skipped=2)
Running unit tests in /tmp/vmdk_ops_unittest/vsan_policy_test.py...
sssss
----------------------------------------------------------------------
Ran 5 tests in 0.000s

OK (skipped=5)
Makefile:145: recipe for target 'test-esx' failed
make[2]: *** [test-esx] Error 1
make[2]: Leaving directory '/home/ajs/go/src/github.com/vmware/docker-volume-vsphere/esx_service'
Makefile:298: recipe for target 'test-esx' failed
make[1]: *** [test-esx] Error 2
make[1]: Leaving directory '/home/ajs/go/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin'
Makefile:33: recipe for target 'all' failed
make: *** [all] Error 2
```

andrewjstone commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -499,27 +499,27 @@ def policy_ls(args):
+        print("Pid: {0}".format(pid))
+        print("Port: {0}".format(get_listening_port(pid)))
+    print("LogConfigFile: {0}".format(log_config.LOG_CONFIG_FILE))
+    printr("LogFile: {0}".format(log_config.LOG_FILE))
```
I just debugged the make all failure and tracked it to this line. Should be `print` instead of `printr`.

govint commented:
1. Rebased on current code in master

2. Ran pylint and fixed those lint-warnings resulting from the new changes.

3. Tested on Python 3.5.1 and 64-bit libs.

4. CI is green at https://ci.vmware.run/vmware/docker-volume-vsphere/839

andrewjstone commented on esx_service/utils/kvESX.py:
```diff
@@ -30,9 +34,10 @@
 DISK_LIB = "/lib/libvmsnapshot.so"
 lib = None
-useSideCarCreate = False
+use_sidecar_create = False
```
Nit: This should probably be in CAPS as well

andrewjstone commented on esx_service/utils/kvESX.py:
```diff
@@ -42,132 +47,161 @@
+   else:
+      return c_uint32(val)
+
+def disk_is_valid(dhandle):
```
Nit: Could shorten this function using `get_uint(0)`

andrewjstone commented on esx_service/vmdk_ops.py:
```diff
@@ -841,23 +853,26 @@ def main():
         if a == '-h':
             usage()
-            return 0
+            return(0)
```
Nit: I find it somewhat weird to wrap a return statement in parens, but it's no big deal

govint replied:
This is enforced by Python 3.5

andrewjstone commented:
LGTM

msterin commented on esx_service/utils/kvESX.py:
```diff
@@ -42,132 +47,161 @@
-    ihandle = c_uint32(0)
+def vol_open_path(volpath):
+    dhandle = get_uint(0)
+    ihandle = get_uint(0)
```
nit: could be `dhandle = ihandle = get_uint(0)`

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -751,13 +756,19 @@ def signal_handler_stop(signalnum, frame):
+def load_vmci():
+   global lib
+
+   logging.info("Loading VMCI server lib.")
```
nit: please add which lib is loaded (64 or 32) to thew log as info

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -841,23 +853,26 @@ def main():
+           logging.exception(msg)
         usage()
-        return 1
+        return(1)
```
actually, I do not think we should use return (val) instead of `return val`. See https://www.python.org/dev/peps/pep-0008/

msterin commented:
LGTM, contingent on addressing a small issue - I think we should use `return 1`, not 'return (1)` style - see PEP8
